[lldb][Windows] WoA HW Watchpoint support in LLDB

omjavaid opened the pull request:
This PR adds support for hardware watchpoints in LLDB for AArch64 Windows 
targets.

Windows does not provide an API to query the number of available hardware 
watchpoints supported by underlying hardware platform. Therefore, current 
implementation supports only a single hardware watchpoint, which has been 
verified on Windows 11 using Microsoft SQ2 and Snapdragon Elite X hardware.                                                                                                                                                                                                                                    

LLDB test suite ninja check-lldb still fails watchpoint-related tests. However, 
tests that do not require more than a single watchpoint pass successfully 
when run individually.

### Known Issues (Just for documentation):

1. **Number of Supported Hardware Breakpoints/Watchpoints:** Windows does not provide the exact number of supported hardware breakpoints or watchpoints. The current implementation guesses this number based on testing how many can be successfully applied on the platform. winnt.h defines ARM64_MAX_WATCHPOINTS = 2 and ARM64_MAX_BREAKPOINTS = 8 however actual number differs.

3. **Initial Stop Behavior:** Hardware breakpoints or watchpoints set on the initial stop do not trigger, even though they are correctly written to the Windows context. They only trigger if set after the main program has started. This issue causes the test suite to fail when it attempts to set hardware breakpoints on the main function before running, as these do not get triggered.

omjavaid committed:
WoA hardware breakpoint/watchpoint support

jasonmolenda commented:
i skimmed the patch, I didn't see anything that concerns me.  The fact that we can't query the number of hardware watchpoints/breakpoints means that this code will misbehave on other platforms where the numbers are different - we might have four hardware watchpoints, but you'll be limited to only setting one.  The hardware might only have 4 hardware breakpoints, so this code will let you set two additional ones which are ignored.  But I don't know what approach would be better here, given what Windows apparently provides.

I'll probably need to touch this new code a tiny bit once it's merged, for https://github.com/llvm/llvm-project/pull/105594 but there's no reason for Omair to do anything about that right now, I'm not ready to land my patches yet.

labath commented on lldb/source/Plugins/Process/Windows/Common/NativeThreadWindows.cpp:
```diff
@@ -178,9 +178,41 @@ Status NativeThreadWindows::RemoveWatchpoint(lldb::addr_t addr) {
 
 Status NativeThreadWindows::SetHardwareBreakpoint(lldb::addr_t addr,
                                                   size_t size) {
+#if defined(__aarch64__) || defined(_M_ARM64)
```
Could we get rid of this ifdef? What would happen if we just let `m_reg_context_up->SetHardwareBreakpoint()` return an error?

omjavaid replied:
Yes I think we can. I ll update this.

labath reviewed (commented):
Seems ok to me.

If I had to guess, I'd say that the initial stop problem is due to us stopping the process "too early" (i.e., in a state where it's not yet fully initialized and ready to accept debugger commands). Maybe there's a different way to stop the process at the first instruction?

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeProcessWindows.cpp:
```diff
@@ -491,24 +491,47 @@ NativeProcessWindows::OnDebugException(bool first_chance,
+        return ExceptionResult::MaskException;
+      } else {
+        const std::vector<ULONG_PTR> &args = record.GetExceptionArguments();
+        if (args.size() >= 2) {
```
What does this function return if this is false?

omjavaid replied:
I believe you're referring to the call below:
reg_ctx.GetWatchpointHitIndex(hw_id, args[1]);
Generally, GetWatchpointHitIndex returns an error only if it fails to read the hardware breakpoint resource information via ReadHardwareDebugInfo. If the watchpoint hit detection is successful, the method returns the hw_watchpoint_id of the corresponding watchpoint. On failure to find watchpoint hit, it returns LLDB_INVALID_INDEX32.

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeRegisterContextWindows_arm64.cpp:
```diff
@@ -143,8 +142,14 @@ NativeRegisterContextWindows::CreateHostNativeRegisterContextWindows(
+    : NativeRegisterContextRegisterInfo(
+          native_thread, CreateRegisterInfoInterface(target_arch)) {
+  // Currently, there is no API to query the maximum supported hardware
+  // breakpoints and watchpoints on Windows. The values set below are based
```
On Windows for ARM64 or in general? I know this is a _arm64 class but just add "on Windows for arm64." in case someone like me is skimming through.

DavidSpickett replied:
>  // on tests conducted on Windows 11 with Snapdragon Elite X hardware.

As in this is what the Windows header had set those defines to, or this is what you were able to use?

I remember this style from my MIPS days but there you could write to the registers and read them back to tell which slots were active. I don't know if that's a viable option here.

omjavaid replied:
I dont think a API exists to query hardware breakpoint or watchpoint resources. On x86_64 as well winnt.h definitions of hardware breakpoint and watchpoints are used.

winnt.h defined the max hardware breakpoints to be 8 however in reality only 6 can be used successfully. Also for watchpoints it defines 2 arm64 hardware watchpoints but in reality only 1 can be used.

DavidSpickett replied:
Does Windows reject writes to the other breakpoints or just silently accepts them and only later you find out they don't work?

I'm wondering if we can read/modify/write the registers and see what sticks.

Of course the timing of that is interesting. We'd need a live process to do those writes in, and perhaps by that time we already needed to know the max breakpoint number for something else to work.

(last time I did this sort of thing it was a jtag debugger so we could inspect registers regardless of the debugee's state)

labath replied:
In principle, there's nothing stopping you from launching a dummy process just for the purpose of testing the debug API. I noticed that this is what linux `strace` does when it tries to determine the host kernel capabilities. Windows makes this slightly trickier as you can't just call `fork` -- you need an actual executable -- but the same principle holds. (I'm not saying we need to do this here, I'm just throwing the idea out there.)

omjavaid replied:
In my testing  I was able to set and read the registers back successfully however the exception which should be generated as a result of setting those registers do not get triggered. So in theory you are able to set and read hardware breakpoint registers back with correct values. But there is some configuration internal to windows which probably does not happen until we land in user code for generating those exceptions.

omjavaid replied:
Also winnt.h defines wvr, wcr, bvr and bcr arrays which has size equal to ARM64_MAX_WATCHPOINTS or ARM64_MAX_BREAKPOINTS. So we should be able to read and write those arrays regardless of whether they eventually generate a watchpoint or breakpoint exception or not.

DavidSpickett replied:
Ok so the kernel structure is always there it just isn't going to write the values to hardware (or it does and the hardware ignores it, but same result either way).

It's a shame because this means an lldb-server built on WoA CpuX won't detect more or less points on CpuY (I think Jason already said this).

I also wouldn't say we need to use Pavel's idea right now. Another option would be to do it based on CPU ID if we have access to that in Windows.

DavidSpickett commented:
Remind me, does lldb support hardware breakpoints for x86 right now either? From the content of this PR, I assume it does not.

(which is fine, I'm just trying to be clear what our starting point is)

DavidSpickett commented:
Relates to https://github.com/llvm/llvm-project/issues/80665

omjavaid commented:
> Remind me, does lldb support hardware breakpoints for x86 right now either? From the content of this PR, I assume it does not.
> 
> (which is fine, I'm just trying to be clear what our starting point is)
yes x86_64 windows does not support hardware breakpoints. However it does support hardware watchpoints.

omjavaid commented:
> Seems ok to me.
> 
> If I had to guess, I'd say that the initial stop problem is due to us stopping the process "too early" (i.e., in a state where it's not yet fully initialized and ready to accept debugger commands). Maybe there's a different way to stop the process at the first instruction?

Yes I think there might be a Windows API launch flag that can let us do early detection of hardware breakpoints and watchpoints. But so far after lots of experimentation I have not been able to find the right set.

DavidSpickett commented:
> Yes I think there might be a Windows API launch flag that can let us do early detection of hardware breakpoints and watchpoints. But so far after lots of experimentation I have not been able to find the right set.

Is it possible this is a bug in Windows itself? Though it seems unlikely given that Visual Studio supports WoA.

I'm not sure about merging this until we know more about that part, at least whether it's a bug or not. Perhaps we (Linaro) can get some information from Microsoft on this?

omjavaid commented:
> > Yes I think there might be a Windows API launch flag that can let us do early detection of hardware breakpoints and watchpoints. But so far after lots of experimentation I have not been able to find the right set.
> 
> Is it possible this is a bug in Windows itself? Though it seems unlikely given that Visual Studio supports WoA.
> 
> I'm not sure about merging this until we know more about that part, at least whether it's a bug or not. Perhaps we (Linaro) can get some information from Microsoft on this?

So further update from my offline (out of LLDB) testing with windows API and setting hardware breakpoints suggest that we might be doing something different on LLDB side during launch or while setting hardware breakpoints/watchpoints that stops us from hitting those exceptions. 

In simple console based debugger I have played with Windows API and set hardware breakpoints and found a good configuration that works.

I will keep debugging this might get to fix it on LLDB side as well. And will talk to our friends at MS at some stage for any clarifications that might be needed.

omjavaid committed:
Remove HW breakpoint support

omjavaid committed:
Run clang-format and fix issues

omjavaid commented:
@DavidSpickett @labath I have updated this PR and dropped hardware breakpoint support for the unresolved "set hw break on main" issue As x86_64 windows also does not support hardware breakpoints (probably for the same reason). 

This PR now only implements HW watchpoint support.

omjavaid committed:
Fix typo: only one watchpoint available at max

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeProcessWindows.cpp:
```diff
@@ -492,23 +492,36 @@ NativeProcessWindows::OnDebugException(bool first_chance,
+        LLDB_LOG(log, "Hit non-loader breakpoint at address {0:x}.",
+                 exception_addr);
+
+        reg_ctx.SetPC(reg_ctx.GetPC() - GetSoftwareBreakpointPCOffset());
```
Does the call to GetSoftwareBreakpointPCOffset replace the comment:
```
// The current PC is AFTER the BP opcode, on all architectures.
```
Want to make sure we don't lose this detail, but if GetSoftwareBreakpointPCOffset handles it then this is fine.

omjavaid replied:
I think it was a typo to miss the comment. I will add it back.

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeProcessWindows.cpp:
```diff
@@ -492,23 +492,36 @@ NativeProcessWindows::OnDebugException(bool first_chance,
+      const auto thread_id = record.GetThreadID();
+
+      if (FindSoftwareBreakpoint(exception_addr)) {
+        LLDB_LOG(log, "Hit non-loader breakpoint at address {0:x}.",
```
What does loader mean in this context? Is it anything like the dynamic linker on Linux? If so I assume this means we hit a breakpoint that wasn't placed by the system, because if it was we should be ignoring it.

omjavaid replied:
"loader breakpoint" as its often called is a system generated breakpoint on executable startup. Loader is Somewhat in similar sequence to how dynamic linker does loading on linux. But on windows system first breakpoint event is something like PTRACE_EVENT_EXEC event. We handle the initial loader breakpoint below in the same function at `if (!initial_stop) {`

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeProcessWindows.cpp:
```diff
@@ -492,23 +492,36 @@ NativeProcessWindows::OnDebugException(bool first_chance,
+        return ExceptionResult::MaskException;
+      } else {
+        const std::vector<ULONG_PTR> &args = record.GetExceptionArguments();
+        if (args.size() >= 2) {
```
Worth adding a comment here to say what the two are, because it's clear that args[1] is an address but you ignore args[0] which is suspicious without context.

omjavaid replied:
Ack.

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeRegisterContextWindows_arm64.cpp:
```diff
@@ -143,8 +142,14 @@ NativeRegisterContextWindows::CreateHostNativeRegisterContextWindows(
+  // breakpoints and watchpoints on Windows. The values set below are based
+  // on tests conducted on Windows 11 with Snapdragon Elite X hardware.
+  m_max_hwp_supported = 1;
+  m_max_hbp_supported = 0;
```
If this patch is only watchpoint support, this should be removed.

omjavaid replied:
Ack.

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeRegisterContextWindows_arm64.cpp:
```diff
@@ -143,8 +142,14 @@ NativeRegisterContextWindows::CreateHostNativeRegisterContextWindows(
+  // Currently, there is no API to query the maximum supported hardware
+  // breakpoints and watchpoints on Windows. The values set below are based
+  // on tests conducted on Windows 11 with Snapdragon Elite X hardware.
+  m_max_hwp_supported = 1;
```
With only 1 useable, are there a lot of tests that don't work?

Which is fine if so, just curious. Enabling watchpoint tests overall doesn't have to be a goal of this PR, if it works at all it'll be an improvement.

omjavaid replied:
Yes. Tests still dont work but following tests pass if we run them individually:

>   lldb-api :: commands/watchpoints/hello_watchpoint/TestMyFirstWatchpoint.py
  lldb-api :: commands/watchpoints/multiple_threads/TestWatchpointMultipleThreads.py
  lldb-api :: commands/watchpoints/step_over_watchpoint/TestStepOverWatchpoint.py  
  lldb-api :: commands/watchpoints/watchpoint_commands/TestWatchpointCommands.py
  lldb-api :: commands/watchpoints/watchpoint_commands/command/TestWatchpointCommandLLDB.py
  lldb-api :: commands/watchpoints/watchpoint_commands/command/TestWatchpointCommandPython.py
lldb-api :: commands/watchpoints/watchpoint_commands/condition/TestWatchpointConditionCmd.py
   lldb-api :: commands/watchpoints/watchpoint_disable/TestWatchpointDisable.py
   lldb-api :: commands/watchpoints/watchpoint_events/TestWatchpointEvents.py
  lldb-api :: commands/watchpoints/watchpoint_on_vectors/TestValueOfVectorVariable.py
  lldb-api :: commands/watchpoints/watchpoint_size/TestWatchpointSizes.py
     lldb-api :: python_api/watchpoint/TestSetWatchpoint.py
  lldb-api :: python_api/watchpoint/TestWatchpointIgnoreCount.py
  lldb-api :: python_api/watchpoint/TestWatchpointIter.py
  lldb-api :: python_api/watchpoint/condition/TestWatchpointConditionAPI.py
  lldb-api :: python_api/watchpoint/watchlocation/TestTargetWatchAddress.py

DavidSpickett reviewed (commented):
So is this X86 and ARM64 or just one?

Also this needs a release note, which is a good place to note the limitations as well.

omjavaid committed:
Fix review comments from David

omjavaid committed:
Add release note for WoA HW Watch support

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeProcessWindows.cpp:
```diff
@@ -492,23 +492,40 @@ NativeProcessWindows::OnDebugException(bool first_chance,
+        // the second contains the virtual address of the accessed data.
+        if (args.size() >= 2) {
+          uint32_t hw_id = LLDB_INVALID_INDEX32;
+          reg_ctx.GetWatchpointHitIndex(hw_id, args[1]);
```
Ideally every error path returns LLDB_INVALID_INDEX32 but you should check that the returned Status is not a failure as well, just in case.

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeProcessWindows.cpp:
```diff
@@ -492,23 +492,40 @@ NativeProcessWindows::OnDebugException(bool first_chance,
+        SetState(eStateStopped, true);
+        return ExceptionResult::MaskException;
+      } else {
+        const std::vector<ULONG_PTR> &args = record.GetExceptionArguments();
```
We know that once we're in this else, it's a hardware break or watch point. But given we don't support hardware breakpoints yet, it can only be a hardware watchpoint.

Assuming that's correct, I would add a comment just after the else to say that.

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeProcessWindows.cpp:
```diff
@@ -492,23 +492,40 @@ NativeProcessWindows::OnDebugException(bool first_chance,
   }
   case DWORD(STATUS_BREAKPOINT):
   case STATUS_WX86_BREAKPOINT:
```
W = something related to WOW64 here right, not W for watchpoint?

And we get this even on arm64, there's no STATUS_WX86_BREAKPOINT? (and certainly you wouldn't get both of those in one process even if there were 2)

omjavaid replied:
yes. Heres the detail from ntstatus.h
//
// MessageId: STATUS_WX86_BREAKPOINT
//
// MessageText:
//
// Exception status code used by Win32 x86 emulation subsystem.
//
#define STATUS_WX86_BREAKPOINT           ((NTSTATUS)0x4000001FL)

DavidSpickett commented on lldb/source/Plugins/Process/Windows/Common/NativeRegisterContextWindows_x86_64.cpp:
```diff
@@ -110,8 +110,8 @@ NativeRegisterContextWindows::CreateHostNativeRegisterContextWindows(
-    : NativeRegisterContextWindows(native_thread,
-                                   CreateRegisterInfoInterface(target_arch)) {}
+    : NativeRegisterContextRegisterInfo(
+          native_thread, CreateRegisterInfoInterface(target_arch)) {}
```
So for i386 and x86_64, you've made these changes to keep them building, but this does not enable writing the watchpoints, correct?

For arm64, you've hooked up the debug register read and write, so there they do work now.

omjavaid replied:
Yes. Correct.
NativeRegisterContextWindows_arm64 class now inherits from both NativeRegisterContextWindows and NativeRegisterContextDBReg_arm64 but both of these classes previously inherited from NativeRegisterContextRegisterInfo non-virtually. This caused multiple instances of NativeRegisterContextRegisterInfo in the hierarchy.
I have now changed NativeRegisterContextWindows definition to inherit NativeRegisterContextRegisterInfo virtually 
which shifts the responsibility of initializing NativeRegisterContextRegisterInfo to the most derived class.
Thats why this and other similar changes have been made.

DavidSpickett commented on llvm/docs/ReleaseNotes.md:
```diff
@@ -575,6 +575,10 @@ Changes to LLDB
+* LLDB now supports hardware watchpoints for AArch64 Windows targets. Windows
+  does not provide API to query the number of supported hardware watchpoints.
+  Therefore current implementation allows only 1 watchpoint, as tested with
+  Windows 11 on the Microsoft SQ2 and Snapdragon Elite X platforms.
```
What about this issue that's causing the test failures, is this something users are going to hit?

Or was that only about breakpoints, it's still in the PR description so I'm not sure.

DavidSpickett replied:
Only thing I can think of is perhaps `check-lldb` at `-j1` manages to start the new processes faster than running the test using lit alone. You could try running one of the tests using lit, in a loop.

But it's not like we don't have watchpoint glitches elsewhere, and if it's a kernel problem, this PR is certainly not going to fix that. So I think we say this is one more reason this feature is experimental.

DavidSpickett commented:
Please update the PR description to reflect what the current code does, or if you want to leave the notes there, move them into an obvious section that you can cut out if/when this gets merged.

omjavaid committed:
Fix unalinged watchpoints

omjavaid commented:
> ### Known Issues (Just for documentation):
> 1.

I have updates the PR description the "Known Issues (Just for documentation):" part will be skipped in the final patch description.

DavidSpickett commented:
I want to be 100% sure of the status of this issue:

> Initial Stop Behavior: Hardware breakpoints or watchpoints set on the initial stop do not trigger, even though they are correctly written to the Windows context. They only trigger if set after the main program has started. This issue causes the test suite to fail when it attempts to set hardware breakpoints on the main function before running, as these do not get triggered.

This is a problem with watchpoints too, right?

I can see an argument that says code breakpoints are much more likely to be placed before the initial stop, so watchpoints are still useful even with this limitation and this PR should still be merged.

But please open an issue for that, describing the problem and as far as you know. why it might be occurring. It will give us something to "mark as duplicate" :)

omjavaid commented:
> This is a problem with watchpoints too, right?

Yes this happens with watchpoints too on both AArch64 and x64 WIndows.

> I can see an argument that says code breakpoints are much more likely to be placed before the initial stop, so watchpoints are still useful even with this limitation and this PR should still be merged.

Yes because LLDB requires a running process to set watchpoints however for breakpoints thats not the case. In most cases user sets breakpoint on main and then set watchpoints which masks this problem altogether.

> But please open an issue for that, describing the problem and as far as you know. why it might be occurring. It will give us something to "mark as duplicate" :)

I have create #125054

DavidSpickett commented:
> In most cases user sets breakpoint on main and then set watchpoints which masks this problem altogether.

Agreed, this isn't ideal but it is an improvement from the current features.

DavidSpickett reviewed (approved):
We're hitting limitations in Windows itself, and 1 watchpoint is more useful than zero, so this LGTM.

To preempt a potential discussion: I think this should not be backported to 20. Unless you've got a very good reason for it. We don't have the most comprehensive testing on Windows and we already have a few caveats here. We might encounter more as time goes on and fixes may be complex.

But, by LLDB 21, maybe we will have ironed those out.

DavidSpickett commented:
Also thanks for tackling this, I know I said once (possibly twice) that I'd do it but kept getting distracted. I know you put a bunch of research time into this too.

dzhidzhoev commented:
@omjavaid 

Apparently, these changes break lldb-remote-linux-win buildbot with the following error:
```
FAILED: tools/lldb/source/Plugins/Process/Windows/Common/CMakeFiles/lldbPluginProcessWindowsCommon.dir/NativeRegisterContextWindows_WoW64.cpp.obj 
ccache C:\PROGRA~1\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1441~1.341\bin\Hostx64\x64\cl.exe  /nologo /TP -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_ENABLE_EXTENDED_ALIGNED_STORAGE -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\source\Plugins\Process\Windows\Common -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\llvm\include -IC:\Python312\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\llvm\..\clang\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\..\clang\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\source -D__OPTIMIZE__ /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -wd4251 -wd4275 -w14062 -we4238 /Gw /O2 /Ob2  -MD   -wd4018 -wd4068 -wd4150 -wd4201 -wd4251 -wd4521 -wd4530 -wd4589  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\lldb\source\Plugins\Process\Windows\Common\CMakeFiles\lldbPluginProcessWindowsCommon.dir\NativeRegisterContextWindows_WoW64.cpp.obj /Fdtools\lldb\source\Plugins\Process\Windows\Common\CMakeFiles\lldbPluginProcessWindowsCommon.dir\lldbPluginProcessWindowsCommon.pdb /FS -c C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common\NativeRegisterContextWindows_WoW64.cpp
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common\NativeRegisterContextWindows_WoW64.cpp(92): error C2280: 'lldb_private::NativeRegisterContextWindows::NativeRegisterContextWindows(void)': attempting to reference a deleted function
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common\NativeRegisterContextWindows.h(29): note: compiler has generated 'lldb_private::NativeRegisterContextWindows::NativeRegisterContextWindows' here
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common\NativeRegisterContextWindows.h(29): note: 'lldb_private::NativeRegisterContextWindows::NativeRegisterContextWindows(void)': function was implicitly deleted because a base class 'lldb_private::NativeRegisterContextRegisterInfo' has either no appropriate default constructor or overload resolution was ambiguous
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins/Process/Utility/NativeRegisterContextRegisterInfo.h(18): note: see declaration of 'lldb_private::NativeRegisterContextRegisterInfo'
65.460 [198/9/5091]Building CXX object tools\lldb\source\Plugins\Process\Windows\Common\CMakeFiles\lldbPluginProcessWindowsCommon.dir\NativeRegisterContextWindows_x86_64.cpp.obj
FAILED: tools/lldb/source/Plugins/Process/Windows/Common/CMakeFiles/lldbPluginProcessWindowsCommon.dir/NativeRegisterContextWindows_x86_64.cpp.obj 
ccache C:\PROGRA~1\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1441~1.341\bin\Hostx64\x64\cl.exe  /nologo /TP -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_ENABLE_EXTENDED_ALIGNED_STORAGE -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\source\Plugins\Process\Windows\Common -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\llvm\include -IC:\Python312\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\llvm\..\clang\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\..\clang\include -IC:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source -IC:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\source -D__OPTIMIZE__ /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -wd4251 -wd4275 -w14062 -we4238 /Gw /O2 /Ob2  -MD   -wd4018 -wd4068 -wd4150 -wd4201 -wd4251 -wd4521 -wd4530 -wd4589  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\lldb\source\Plugins\Process\Windows\Common\CMakeFiles\lldbPluginProcessWindowsCommon.dir\NativeRegisterContextWindows_x86_64.cpp.obj /Fdtools\lldb\source\Plugins\Process\Windows\Common\CMakeFiles\lldbPluginProcessWindowsCommon.dir\lldbPluginProcessWindowsCommon.pdb /FS -c C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common\NativeRegisterContextWindows_x86_64.cpp
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common\NativeRegisterContextWindows_x86_64.cpp(114): error C2280: 'lldb_private::NativeRegisterContextWindows::NativeRegisterContextWindows(void)': attempting to reference a deleted function
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common\NativeRegisterContextWindows.h(29): note: compiler has generated 'lldb_private::NativeRegisterContextWindows::NativeRegisterContextWindows' here
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins\Process\Windows\Common\NativeRegisterContextWindows.h(29): note: 'lldb_private::NativeRegisterContextWindows::NativeRegisterContextWindows(void)': function was implicitly deleted because a base class 'lldb_private::NativeRegisterContextRegisterInfo' has either no appropriate default constructor or overload resolution was ambiguous
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\source\Plugins/Process/Utility/NativeRegisterContextRegisterInfo.h(18): note: see declaration of 'lldb_private::NativeRegisterContextRegisterInfo'

```

https://lab.llvm.org/buildbot/#/builders/197/builds/1447/steps/8/logs/stdio

https://lab.llvm.org/buildbot/#/builders/197/builds/1447

Could you take care of it?

omjavaid commented:
> https://lab.llvm.org/buildbot/#/builders/197/builds/1447/steps/8/logs/stdio
> 
> https://lab.llvm.org/buildbot/#/builders/197/builds/1447
> 
> Could you take care of it?

@dzhidzhoev I am taking a look. Thanks for spotting this.
